crates/modelardb_storage/src/query/metadata_table.rs

```diff
@@ -0,0 +1,83 @@
+/* Copyright 2024 The ModelarDB Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Implementation of [`MetadataTable`] which allows metadata tables to be queried through Apache
+//! DataFusion. It wraps a [`DeltaTable`] and forwards most method calls to it. However, for
+//! [`TableProvider::scan()`] it updates the [`DeltaTable`] to the latest version.
+
+use std::{any::Any, sync::Arc};
+
+use datafusion::catalog::Session;
+use datafusion::datasource::{TableProvider, TableType};
+use datafusion::error::{DataFusionError, Result as DataFusionResult};
+use datafusion::logical_expr::Expr;
+use datafusion::physical_plan::ExecutionPlan;
+use deltalake::{arrow::datatypes::SchemaRef, DeltaTable};
+use tonic::async_trait;
+
+/// A queryable representation of a metadata table. [`MetadataTable`] wraps the [`TableProvider`] of
+/// [`DeltaTable`] and passes most methods calls directly to it. Thus, it can be registered with
+/// Apache Arrow DataFusion. The only difference from [`DeltaTable`] is that `delta_table` is
+/// updated to the latest snapshot when accessed.
+pub(crate) struct MetadataTable {
```

Comment:
Unclear why `NormalTable` is not used for both metadata tables and normal tables as both seems to be Apache Parquet files in a Delta Lake table. Thus, it seems possible that one `TableProvider` type could be used for both.

Reply:
I think it is fine to have a little duplicated code for a cleaner interface and to avoid having to pass a `DataSink` for metadata tables. However, I have added a note to https://github.com/ModelarData/ModelarDB-RS/issues/258 to look into this further.